submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,427 @@
+cloud (assuming they are in the same ClusterSet) as well as an IPSec cable between clusters in different
+clouds.
+
+In addition to supporting varying cable types between clusters in a ClusterSet, the user will also select
+which clusters to connect or not connect.
+
+> **_NOTE:_** The proposal assumes that connection between two clusters is symmetrical. If Cluster A is
+connected to Cluster B via a VXLAN cable type - then Cluster B is connected to Cluster A via a VXLAN cable
+type and not another cable type.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect
+is eventually realized into the cluster's desired state. To extend the Kubernetes API, one could either use
+the generic Kubernetes `ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each
+but a CRD seems more suitable for this enhancement.
+
+This enhancement will propose a single API for finer-grained connectivity and multiple cable driver support.
+
+### Finer-grained connectivity and multiple cable driver API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    ConnectionPolicies []ClusterConnectionPolicySpec `json:"spec"`
+}
+
+type ClusterConnectionPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+   // Priority identifies the priority for the connection policy. Higher priorities are applied first.
+    Priority  *int  `json:"priority,omitempty"`
+
+    // ConnectionPolicy identifies the default connection policy if no policies are found: to connect or no-connect.
+    ConnectionPolicy *string `json:"connectionPolicy,omitempty"`
+
+    CableDriverConfigs []CableDriverSpec `json:"spec"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    CableOptions *v1.ConfigMap `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectionPolicies` CR defines the criteria for selecting which cluster pairs/subsets can connect and
+associated optional configuration options to use for each cluster connection. To resolve the connectivity policy
+for a remote cluster's Endpoint, Submariner would:  
```

Comment:
I've been thinking about the order of processing here and the complexity added by having a priority... 

From a practical POV of what would be preferable from an implementation - is to remove the priority altogether. There's an onus on the user to put together a set of policies that describe the desirable end state for the cluster connections.

If priorities are gone then it's also important to have a default policy which will be called `default` in order to distinguish it from everything else.

So in order to process these policies (looking at this purely from a connectivity POV and excluding the cable driver for the minute), the GW could build a connection map[string]string , where the key is a cluster identifier or potentially even an endpoint identifier and the value is connect or no-connect. On the first pass the default policy is applied to all values in the map. Establishing a connectivity list that's either a permit-list or a deny-list based on the default policy. 

The next thing the GW should do is parse the rest of the policies - where it finds a policy whose left selectors match it, it establishes the list of clusters that match the right selectors and updates the connectivity map with what's specified in the policy (this info could be pushed as a ConfigMap to the local K8s API so that it's available for other GWs - or just let the GWs do the relevant configuration when they come up). 

When the time comes to actually install the cable - the connectivity map is consulted... 

or something along those lines...

Reply:
For cable driver connection policies, we would need priorities in some manner to avoid ambiguity should a cluster match multiple policies. For connectivity, if we go with a simpler approach where a policy only species what to connect then it's a simple binary decision so priorities aren't needed, ie we need only search for one policy that matches.  But if we add the `connect`/`no-connect` option, then I think we do need priorities to resolve a case where a cluster matches both a `connect` and a `no-connect` policy.